src/gateway/LastTxSyncer.ts

```diff
@@ -0,0 +1,84 @@
+import {Mutex, MutexInterface} from "async-mutex";
+import {Knex} from "knex";
+import {LoggerFactory} from "warp-contracts";
+
+export class LastTxSync {
+
+  private readonly logger = LoggerFactory.INST.create(LastTxSync.name);
+
+  constructor(private readonly gatewayDb: Knex) {
+  }
+
+  // a map from contractTxId to Mutex
+  private readonly contractsMutex = new Map<string, Mutex>();
+
+  // a map from contractTxId to sortKey - i.e. the 'previous' sortKey
+  private readonly lastSortKey = new Map<string, string | null>();
+
+  // a global mutex used to acquire contract mutex from contract->mutex map
+  // for simplicity we could use only this global mutex, but this would greatly reduce tps
+  // - we could process only one transaction at any time.
+  // With per contract mutex - we can process one transaction PER contract.
+  private readonly mainMutex = new Mutex();
+
+  async acquireMutex(contractTxId: string): Promise<MutexInterface.Releaser> {
+    // note: operation of acquiring per-contract mutex is not atomic
+    // - and therefore needs to be synchronized globally - via 'mainMutex'
+    // without this synchronization it could happen that two requests would
+    // create and acquire two independent 'contract' mutexes.
+    const mainRelease = await this.mainMutex.acquire();
```

Comment:
it will be catched in https://github.com/warp-contracts/gateway/pull/110/files#diff-925002587e7eba50a7804e4021c11fc908454e00be475613a591a693fdb4723fR165